dango/testing/tests/factory.rs

```diff
@@ -334,85 +352,51 @@ fn update_key() {
         .should_succeed_and_equal(btree_map! { key_hash => pk });
 }
 
-/// A malicious block builder detects a register user transaction, inserts a new,
-/// false transaction that substitutes the legitimate transaction's secret,
-/// key, or key hash. Should fail because the derived deposit address won't match.
-#[test_case(
-    Some(5),
-    None,
-    None;
-    "false secret"
-)]
-#[test_case(
-    None,
-    Some(Key::Secp256k1(ByteArray::from([0; 33]))),
-    None;
-    "false key"
-)]
-#[test_case(
-    None,
-    None,
-    Some(Hash256::from_inner([0; 32]));
-    "false key hash"
-)]
-fn false_factory_tx(
-    false_secret: Option<u32>,
-    false_key: Option<Key>,
-    false_key_hash: Option<Hash256>,
-) {
+/// A malicious block builder detects a register user transaction, could try to,
+/// false transaction that substitutes the legitimate transaction's username.
+/// Should fail because the signature won't be valid.
+#[test]
+fn malicious_register_user() {
     let (mut suite, _, codes, contracts) = setup_test_naive();
 
     // User makes the deposit normally.
     let user = TestAccount::new_random("user").predict_address(
         contracts.account_factory,
-        2,
         codes.account_spot.to_bytes().hash256(),
         true,
     );
 
-    let secret = false_secret.unwrap_or(2);
-    let key = false_key.unwrap_or(user.first_key());
-    let key_hash = false_key_hash.unwrap_or(user.first_key_hash());
+    let attacker = TestAccount::new_random("attacker").predict_address(
+        contracts.account_factory,
+        codes.account_spot.to_bytes().hash256(),
+        true,
+    );
 
     // A malicious block builder sends a register user tx with falsified
-    // secret, key, or key hash.
-    //
-    // Should fail with "data not found" error, because it be different deposit
-    // address for which no deposit is found.
+    // username signature.
     //
-    // We test with `FinalizedBlock` here instead of with `CheckTx`, because a
-    // malicious block builder can bypass mempool check.
+    // Should fail with "unauthentiec" error, because it signing with the wrong
+    // key for the username.
     suite
         .send_message(
             &mut Factory::new(contracts.account_factory),
             Message::execute(
                 contracts.account_factory,
                 &account_factory::ExecuteMsg::RegisterUser {
-                    username: user.username.clone(),
-                    secret,
-                    key,
-                    key_hash,
+                    data: NewUserSalt {
+                        username: user.username.clone(),
+                        key: user.first_key(),
+                        key_hash: user.first_key_hash(),
+                    },
+                    signature: attacker
+                        .sign_arbitrary(json!({
+                            "username": "random_username",
+                        }))
+                        .unwrap(),
```

Comment:
@j0nl1 I think the more reasonable test case is: the attacker changes the username in the message, but keeps the user's signature.

Reply:
Actually let me add tests for all possible cases.